### PR TITLE
feat(UI): take back to learn map after finishing a block

### LINKED ISSUE #49170: Take campers back to map after finishing a block
Just opening this up for discussion since it came to mind. I was testing a PR and submitted a new RWD project and it took me right to the first step of the next block. I'm wondering if we should take campers back to the map so they can see their checkmark and then they have to click on the next thing to do. It feels like that might be a little better to me.

### PR DESCRIPTION
**Pull request by DEVelooper29**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #49170 

Now, campers are taken back to the map after finishing a block instead of directly starting the next block step-1 in Responsive Web Design Certification.

### CODE REVIEW

**Comment by DEVelooper29** [comment]
Hey @moT01, Can you please review this and let me know if any changes are to be done?

**Comment by raisedadead** [comment]
Please rebase and make sure the tests are passing. Thanks.

**Comment by DEVelooper29** [comment]
> Please rebase and make sure the tests are passing. Thanks.

Done sir. Now all tests are passing and also I've done the rebase.

**Comment by DEVelooper29** [comment]
> Please rebase and make sure the tests are passing. Thanks.
I have successfully done the rebase as you mentioned and also all tests are passing except Firefox.

**Comment by raisedadead** [comment]
You would need to update the test for the changed functionality. We have outlined working with tests in our guide. Please reach out to the Discord chat for more help :)

**Comment by raisedadead** [comment]
This PR now has irrelevant changes. 

I am unsure if you are trying some different method, but this project has specific guidelines about rebasing. There are detailed in the guidelines we linked earlier: https://contribute.freecodecamp.org/#/how-to-open-a-pull-request?id=conflicts-on-a-pull-request

**Comment by DEVelooper29** [comment]
> This PR now has irrelevant changes.
> 
> I am unsure if you are trying some different method, but this project has specific guidelines about rebasing. There are detailed in the guidelines we linked earlier: https://contribute.freecodecamp.org/#/how-to-open-a-pull-request?id=conflicts-on-a-pull-request

I tried running all the commands following the document that you sent. But still, the tests are failing because of the files that were edited after syncing fork. Please help me out.

File: api/src/server.test.ts [hunk 1]
```diff
@@ -1,4 +1,7 @@
+<<<<<<< HEAD
 /* eslint-disable @typescript-eslint/no-unsafe-member-access */
+=======
+>>>>>>> 1416188439 (added feature: take back to learn map)
```
Comment (naomi-lgbt) [inline api/src/server.test.ts:4]:
```suggestion
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
```

Merge conflict markers got left behind

**Review by naomi-lgbt — DISMISSED** [review]

**Comment by DEVelooper29** [comment]
I have solved the issue and also all the tests are passing. Kindly review and please let me know any suggestions/changes required. And sorry for the errors that I had made. This being my first contribution, I faced some errors. I'll be more careful from next time onwards.

**Review by moT01 — DISMISSED** [review]
Thanks for working on this @DEVelooper29 👍

I don't particularly like how we are doing some of this - checking if the path includes the superBlock, for instance - and (semi) hard-coding the `pathToNavigateTo` e.g. `/learn/${superBlock}/#${superBlock}-projects`. Some of this was pre-existing, and I think you were just following that. But what I think what we should do is add info from the next challenge to the challenge meta. So we could then directly compare the current challenge `superblock` with the next challenge `superblock` - and same for with the blocks - and add other useful data if needed. Some of this feels like we are guessing what that info is - which takes me to this path `/learn/2022/responsive-web-design/#2022/responsive-web-design-projects` after I finished the last project on the new RWD (that was pre-existing).

Functionally, this seems to work, but it's a little confusing and I don't like all this hard coded stuff and comparisons against paths.

<hr>

I wrote some code while looking at this PR I think we can use to make all this easier:

<details><summary>challenge-page-creator.js</summary>
At the bottom of the <a target="_blank" href="https://github.com/freeCodeCamp/freeCodeCamp/blob/main/client/utils/gatsby/challenge-page-creator.js">challenge-page-creator.js</a> file, replace what's there with this:

```js
function getNextChallengeMeta(_node, index, nodeArray) {
  const next = nodeArray[index + 1];
  if (next) {
    const { superBlock, block } = next.node.challenge;
    return {
      superBlock,
      block,
      blockSlug: createBlockSlug(_node)
    }
  }
  return null;
}

function createBlockSlug(_node) {
  if (_node) {
    const { block, fields: { slug }} = _node;
    const re = new RegExp(`${block}.*`);
    return slug.replace(re, `#${block}`);
  }
  return null;
}

exports.createChallengePages = function (createPage) {
  return function ({ node: { challenge } }, index, allChallengeEdges) {
    const {
      certification,
      superBlock,
      block,
      fields: { slug },
      required = [],
      template,
      challengeType,
      id
    } = challenge;
    // TODO: challengeType === 7 and isPrivate are the same, right? If so, we
    // should remove one of them.

    createPage({
      path: slug,
      component: getTemplateComponent(challengeType),
      context: {
        challengeMeta: {
          blockSlug: createBlockSlug(challenge),
          certification,
          superBlock,
          block,
          isFirstStep: getIsFirstStep(challenge, index, allChallengeEdges),
          template,
          required,
          nextChallengeMeta: getNextChallengeMeta(
            challenge,
            index,
            allChallengeEdges
          ),
          nextChallengePath: getNextChallengePath(
            challenge,
            index,
            allChallengeEdges
          ),
          prevChallengePath: getPrevChallengePath(
            challenge,
            index,
            allChallengeEdges
          ),
          id
        },
        projectPreview: getProjectPreviewConfig(challenge, allChallengeEdges),
        slug
      }
    });
  };
};
```
I created the two functions at the top and added the `nextChallengeMeta` and `blockSlug` in the `createPage` function. You will need to rebuild the client for these changes to work.
</details>

<details><summary>redux/index.js</summary>
Then in the <a target="_blank" href="https://github.com/freeCodeCamp/freeCodeCamp/blob/main/client/src/redux/index.js">redux/index.js</a> file, add the `nextChallengeMeta` to the initial state:

```js
const initialState = {
  ...
  challengeMeta: {
    superBlock: '',
    block: '',
    id: '',
    nextChallengeMeta: {
      superBlock: '',
      block: '',
      blockSlug: '/'
    },
    nextChallengePath: '/',
    prevChallengePath: '/',
    challengeType: -1
  },
  ...
};
```
</details>

After adding these two changes, you should be able to pull the `superblock` and `block` from both the current challenge and next challenge to compare - and use the `blockSlug` as the `pathToNavigateTo`. I didn't get that far, but I think it should work. Any interest in giving that a try @DEVelooper29? If you need any help, let us know.

**Comment by DEVelooper29** [comment]
> 

Thank You so much. I'll do this.

**Comment by CallmeHongmaybe** [comment]
So @DEVelooper29 do you need any help with the PR still ?

**Comment by moT01** [comment]
I made a few updates to this - it should be good to go.

**Comment by naomi-lgbt** [comment]
Do we want to take the camper back to the project they just completed, or to the next project?

**Comment by moT01** [comment]
Either way works. I like both for their own reasons. Back to the project they just completed means they can see their checkmarks (also, that's how this PR is atm). To the next project is where the next challenge is, so maybe that's better - don't know.

**Review by Sembauke — APPROVED** [review]

**Review by naomi-lgbt — APPROVED** [review]

**Comment by naomi-lgbt** [comment]
Thank you for your contribution to the page! 👍
We are happy to accept these changes and look forward to future contributions. 🎉

**Pull request merged**